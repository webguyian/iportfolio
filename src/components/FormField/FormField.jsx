import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const FormField = props => {
  const {
    className,
    id,
    label,
    disabled,
    onChange,
    type,
    value,
    ...otherProps
  } = props;
  const isTextarea = type === 'textarea';
  const baseClass = 'ui-form-field';
  const handleChange = event => {
    const newValue = event.target.value;

    onChange(id, newValue, event);
  };

  if (isTextarea) {
    return (
      <div className={classNames(baseClass, className)}>
        <textarea
          className={`${baseClass}-textarea`}
          id={id}
          value={value}
          onChange={handleChange}
          {...otherProps}
        />
      </div>
    );
  }

  return (
    <div className={baseClass}>
      <label className={`${baseClass}-label`} htmlFor={id}>
        {label}
      </label>
      <input
        className={`${baseClass}-input`}
        id={id}
        disabled={disabled}
        autoComplete="off"
        onChange={handleChange}
        readOnly={Boolean(!onChange)}
        type={type}
        value={value}
        {...otherProps}
      />
    </div>
  );
};

FormField.propTypes = {
  className: PropTypes.string,
  id: PropTypes.string.isRequired,
  label: PropTypes.string,
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
  type: PropTypes.string,
  value: PropTypes.string
};

FormField.defaultProps = {
  disabled: false,
  type: 'text',
  value: ''
};

export default FormField;
