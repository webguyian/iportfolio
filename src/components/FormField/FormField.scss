.ui-form-field {
  display: flex;
  padding: 8px;

  &-label,
  &-input,
  &-textarea {
    @include type-body;
    color: $white;
  }

  &-label {
    margin-right: 8px;
  }

  &-input,
  &-textarea {
    width: 100%;
    border: 0;
    background: none;
    appearance: none;
  }

  &-textarea {
    resize: none;
  }
}
