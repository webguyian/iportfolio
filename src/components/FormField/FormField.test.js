import React from 'react';
import { create } from 'react-test-renderer';

import FormField from './FormField';

describe('<FormField />', () => {
  const props = {
    id: 'field',
    label: 'Field label',
    onChange: jest.fn(),
    value: ''
  };

  it('renders correctly', () => {
    const component = create(<FormField {...props} />);

    expect(component).toMatchSnapshot();
  });

  it('renders correctly as textarea', () => {
    const component = create(<FormField {...props} type="textarea" />);

    expect(component).toMatchSnapshot();
  });

  it('renders correctly when disabled', () => {
    const component = create(<FormField {...props} disabled />);

    expect(component).toMatchSnapshot();
  });

  it('renders correctly without onChange prop', () => {
    const component = create(<FormField {...props} onChange={undefined} />);

    expect(component).toMatchSnapshot();
  });

  it('handles onChange event', () => {
    const component = create(<FormField {...props} />).root;
    const input = component.findByType('input');
    const mockEvent = {
      target: {
        value: 'test'
      }
    };

    expect(props.onChange).not.toHaveBeenCalled();
    input.props.onChange(mockEvent);
    expect(props.onChange).toHaveBeenCalledWith(props.id, 'test', mockEvent);
  });
});
