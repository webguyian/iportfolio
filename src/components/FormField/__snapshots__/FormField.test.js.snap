// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormField /> renders correctly 1`] = `
<div
  className="ui-form-field"
>
  <label
    className="ui-form-field-label"
    htmlFor="field"
  >
    Field label
  </label>
  <input
    autoComplete="off"
    className="ui-form-field-input"
    disabled={false}
    id="field"
    onChange={[Function]}
    readOnly={false}
    type="text"
    value=""
  />
</div>
`;

exports[`<FormField /> renders correctly as textarea 1`] = `
<div
  className="ui-form-field"
>
  <textarea
    className="ui-form-field-textarea"
    id="field"
    onChange={[Function]}
    value=""
  />
</div>
`;

exports[`<FormField /> renders correctly when disabled 1`] = `
<div
  className="ui-form-field"
>
  <label
    className="ui-form-field-label"
    htmlFor="field"
  >
    Field label
  </label>
  <input
    autoComplete="off"
    className="ui-form-field-input"
    disabled={true}
    id="field"
    onChange={[Function]}
    readOnly={false}
    type="text"
    value=""
  />
</div>
`;

exports[`<FormField /> renders correctly without onChange prop 1`] = `
<div
  className="ui-form-field"
>
  <label
    className="ui-form-field-label"
    htmlFor="field"
  >
    Field label
  </label>
  <input
    autoComplete="off"
    className="ui-form-field-input"
    disabled={false}
    id="field"
    onChange={[Function]}
    readOnly={true}
    type="text"
    value=""
  />
</div>
`;
